### PR TITLE
Test that we are inline with the BSON Object ID spec

### PR DESCRIPTION
**Pull request by nick96**
I've added the spec and have done a pass on checking we're inline with it. I've (and claude)
also added tests to make sure we stay inline with it. There were a few places
where we weren't compliant with the spec, see individual commits for details.

### CODE REVIEW

File: crates/object_id/src/lib.rs [hunk 1]
```diff
@@ -49,14 +52,19 @@ impl ObjectId {
     /// Generate an, optionally unique, object ID for the given time.
     pub fn from_time(t: i64, unique: bool) -> Self {
         let counter = if unique {
-            let counter = COUNTER.fetch_add(1, Ordering::SeqCst);
+            // We don't care that there is an ordering between threads, just
+            // that each ID gets a unique value.
+            let counter = COUNTER.fetch_add(1, Ordering::Relaxed) & COUNTER_MAX;
```
Comment (shannoncole) [inline crates/object_id/src/lib.rs:57]: I don't understand this use of `&`; does it work? It seems we'll wrap and then mask out any bits that aren't in `COUNTER_MAX`; is that what we want?

File: crates/object_id/src/lib.rs [hunk 2]
```diff
@@ -49,14 +52,19 @@ impl ObjectId {
...
             [bs[1], bs[2], bs[3]]
         } else {
             [0, 0, 0]
         };
 
-        // TODO: Handle overflow
+        // Ensure the timestamp is 4 bytes. For timestamps far in the future
+        // this will truncate them but it's how the spec is defined. Note: if
+        // `t` is bigger than u32::MAX, the value will be truncated. This is
+        // expected.
```
Comment (shannoncole) [inline crates/object_id/src/lib.rs:67]: What about negative time-stamps?

File: crates/object_id/src/lib.rs (hunk 1 above)
Comment (nick96) [reply crates/object_id/src/lib.rs:57]: Yeah it does work. It just zeros out bits on anything larger than `COUNTER_MAX` causing it to wrap at 24 bits

File: crates/object_id/src/lib.rs [hunk 3]
```diff
@@ -134,3 +137,110 @@ impl Default for ObjectId {
...
+        assert_eq!(id.timestamp(), 0x7FFFFFFF);
+
+        // 0x80000000: Jan 19th, 2038 03:14:08 UTC
+        let id = ObjectId::from_time(0x80000000u32 as i64, false);
+        assert_eq!(id.timestamp(), 0x80000000u32 as i64);
+
+        // 0xFFFFFFFF: Feb 7th, 2106 06:28:15 UTC
+        let id = ObjectId::from_time(0xFFFFFFFFu32 as i64, false);
+        assert_eq!(id.timestamp(), 0xFFFFFFFFu32 as i64);
+    }
```
Comment (shannoncole) [inline crates/object_id/src/lib.rs:225]: can we also test `i64` values that don't fit in `u32`? ie: too big, negative

File: specifications/objectid.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
+# BSON ObjectID
```
Comment (shannoncole) [inline specifications/objectid.md:1]: oh good idea adding this!

File: crates/object_id/src/lib.rs (hunk 2 above)
Comment (nick96) [reply crates/object_id/src/lib.rs:67]: It works-ish. It doesn't panic and you'll more than likely get a unique ID but you won't be able to get the same timestamp back. It's called out in the spec that this is explicitly not handled because MongoDB didn't exist before 1970.

**Comment by shannoncole** [comment]
new tests look good! brew r+ 👍
